### PR TITLE
Mchang/add congress

### PR DESCRIPTION
**Pull request by maximilianchang**

### CODE REVIEW

File: public/contracts/BountyDistribution.sol [hunk 1]
```diff
@@ -1,37 +1,183 @@
...
+        mapping(uint256 => uint256) indexes; 
+        uint minExecutionDate; 
+
+        uint numAnswers; 
         bool settled;
         bool isValue; // checks for existence of key
     }
 
+    struct Answer {
+        address answerAddr;
```
Comment (shu-claire) [inline public/contracts/BountyDistribution.sol]: answerer

File: public/contracts/BountyDistribution.sol [hunk 2]
```diff
@@ -1,37 +1,183 @@
...
+    constructor() public {
+        membership[msg.sender] = true; 
+    }
+
+    // Membership Functions
+    function newProposal(address recipient, bool addAddress) onlyMembers public {
+        if (addAddress) {
+            require(!addProposals[recipient].isValid || !addProposals[recipient].open);
+        } else {
+            require(!removeProposals[recipient].isValid || !removeProposals[recipient].open);
```
Comment (shu-claire) [inline public/contracts/BountyDistribution.sol]: require member to be a part of the organization already in order for them to be removed

File: public/app/js/newsfeed-view.js [hunk 3]
```diff
@@ -60,10 +58,9 @@ function submitQuestion() {
     time_exp: q_time_exp
   }
 
-  PostModel.add(question_data, function(body) {
-    console.log(JSON.parse(body)); 
-    collectBounty(JSON.parse(body).qHash, question_data.bounty); 
-  })
+  collectBounty(JSON.parse(body).qHash, question_data.bounty, {
```
[the remark this answers is not in the document]
Comment (maximilianchang) [reply public/app/js/newsfeed-view.js]: Yeah, I changed the parameters of collectBounty, so we need to change this .

File: public/contracts/BountyDistribution.sol [hunk 4]
```diff
@@ -1,37 +1,183 @@
...
-    address contractCreator;
 
-    function BountyDistribution() public { // constructor
-        contractCreator = msg.sender;
+    constructor() public {
+        membership[msg.sender] = true; 
+    }
+
+    // Membership Functions
+    function newProposal(address recipient, bool addAddress) onlyMembers public {
```
[the remark this answers is not in the document]
Comment (maximilianchang) [reply public/contracts/BountyDistribution.sol]: changed to proposedMember

File: public/contracts/BountyDistribution.sol [hunk 5]
```diff
@@ -1,37 +1,183 @@
...
+                currentResult: 0,
+                open: true,
+                isValid: true
+            });
+        }
     }
 
-    function distributeBounty(address _answerer, uint256 _aHash, uint256 _qHash) public {
+    function vote(address recipient, bool add, bool support) onlyMembers public {
+        MemberProposal storage p;
```
[the remark this answers is not in the document]
Comment (maximilianchang) [reply public/contracts/BountyDistribution.sol]: mmm...yeah, i see it in my linting. seems like their linter isn't very robust. if you can deploy it regardless, let's leave it for now.

File: public/contracts/BountyDistribution.sol [hunk 6]
```diff
@@ -1,37 +1,183 @@
...
+            p = addProposals[recipient];
+        } else {
+            p = removeProposals[recipient]; 
+        }
+        require (p.isValid); 
+
+        require(!p.voted[msg.sender]); 
+        p.numberOfVotes++; 
+        if (support) {
+            p.currentResult++;
```
[the remark this answers is not in the document]
Comment (maximilianchang) [reply public/contracts/BountyDistribution.sol]: Good point. Right now, in the struct, we are recording "numberOfVotes"; we aren't using it right now, but if we wanted to know we could leverage this to do what you propose.

File: public/contracts/BountyDistribution.sol [hunk 7]
```diff
@@ -1,37 +1,183 @@
...
+            answerAddr: msg.sender,
+            answerHash: _aHash,
+            numUpvotes: 0,
+            isValue: true
+        }); 
+        questions[_qHash].indexes[_aHash] = questions[_qHash].numAnswers; 
+        questions[_qHash].numAnswers++; 
+    }
+
+    function addUpvote(uint256 _qHash, uint256 _aHash) onlyMembers public {
```
[the remark this answers is not in the document]
Comment (maximilianchang) [reply public/contracts/BountyDistribution.sol]: changed

File: public/contracts/BountyDistribution.sol [hunk 8]
```diff
@@ -1,37 +1,183 @@
...
+        MemberProposal storage p; 
+        if (add) { 
+            p = addProposals[recipient];
+        } else {
+            p = removeProposals[recipient]; 
+        }
+        require (p.isValid); 
+
+        require(!p.voted[msg.sender]); 
+        p.numberOfVotes++;
```
Comment (shu-claire) [inline public/contracts/BountyDistribution.sol]: needs to add voter address to the `voted` mapping

File: public/contracts/BountyDistribution.sol [hunk 9]
```diff
@@ -1,37 +1,183 @@
...
+
+    // Congress Variables/Structs
+    mapping (address => bool) membership; 
+    mapping (address => MemberProposal) addProposals; 
+    mapping (address => MemberProposal) removeProposals; 
+
+    uint numMembers; 
+    uint debatingPeriodInMinutes = 600; 
+
+    struct MemberProposal {
```
[the remark this answers is not in the document]
Comment (maximilianchang) [reply public/contracts/BountyDistribution.sol]: i initially had it structured like this, but it made the code a little bit more messy. I'm going to ignore this for now, but it's a valid point and we can revisit it later.

File: public/contracts/BountyDistribution.sol (hunk 8 above)
Comment (maximilianchang) [reply public/contracts/BountyDistribution.sol]: resolved.

File: public/contracts/BountyDistribution.sol [hunk 10]
```diff
@@ -1,37 +1,183 @@
...
+        }
     }
 
-    function distributeBounty(address _answerer, uint256 _aHash, uint256 _qHash) public {
+    function vote(address recipient, bool add, bool support) onlyMembers public {
+        MemberProposal storage p; 
+        if (add) { 
+            p = addProposals[recipient];
+        } else {
+            p = removeProposals[recipient];
```
[the remark this answers is not in the document]
Comment (maximilianchang) [reply public/contracts/BountyDistribution.sol]: see Peter's earlier comment

File: public/contracts/BountyDistribution.sol [hunk 11]
```diff
@@ -1,37 +1,183 @@
...
+        }
+    }
+
+    // Q&A Functions
+    function addAnswer(uint256 _qHash, uint256 _aHash) onlyMembers public {
         require(questions[_qHash].isValue && !questions[_qHash].settled);
-        _answerer.transfer(questions[_qHash].bounty);
-        questions[_qHash].answerHash = _aHash;
-        questions[_qHash].settled = true;
+        questions[_qHash].answers[questions[_qHash].numAnswers] = Answer({
```
[the remark this answers is not in the document]
Comment (maximilianchang) [reply public/contracts/BountyDistribution.sol]: don't think there is one. fixed.

File: public/contracts/BountyDistribution.sol [hunk 12]
```diff
@@ -1,37 +1,183 @@
...
+        p.numberOfVotes++; 
+        if (support) {
+            p.currentResult++; 
+        } else {
+            p.currentResult--; 
+        }
+    }
+
+    function executeProposal(address recipient, bool add) public {
+        MemberProposal storage p;
```
[the remark this answers is not in the document]
Comment (maximilianchang) [reply public/contracts/BountyDistribution.sol]: not sure either, tbh

**Pull request merged**